scripts/o3de/o3de/compatibility.py

```diff
@@ -0,0 +1,245 @@
+            return set()
+        else:
+            incompatible_objects.update(incompatible_apis)
+
+    return incompatible_objects
+
+
+def get_incompatible_gem_version_specifiers(gem_version_specifier_list:list, all_gems_json_data:dict) -> set:
+    """
+    Returns a set of gem version specifiers that are not compatible with the gem's provided
+    If a gem_version_specifier_list entry only has a gem name, it is assumed compatible with every gem version with that name.
+    :param gem_version_specifier_list: a list of gem names and (optional)version specifiers
+    :param all_gems_json_data: json data of all gems to use for compatibility checks
+    """
+    if not gem_version_specifier_list:
+        return set()
+
+    if not all_gems_json_data:
+        # no gems are available
+        return set(gem_version_specifier_list)
+
+    incompatible_gem_version_specifiers = set()
+    for gem_version_specifier in gem_version_specifier_list:
+        gem_name, version_specifier = utils.get_object_name_and_optional_version_specifier(gem_version_specifier)
+        if not gem_name in all_gems_json_data:
+            incompatible_gem_version_specifiers.add(f"{gem_version_specifier} (missing dependency)")
+            continue
+
+        if not version_specifier:
+            # when no version specifier is provided we assume compatibility with any version
+            continue
+        
+        gem_version = all_gems_json_data[gem_name].get('version','')
+        if gem_version and not has_compatible_version([gem_version_specifier], gem_name, gem_version):
+            incompatible_gem_version_specifiers.add(f"{gem_version_specifier} (different version found ${gem_version})")
+            continue
+
+    return incompatible_gem_version_specifiers
+
+
+def has_compatible_name(name_and_version_specifier_list:list, object_name:str) -> bool:
+    """
+    Returns True if the object_name matches an entry in the name_and_version_specifier_list
+    :param name_and_version_specifier_list: a list of names and (optional)version specifiers
+    :param object_name: the object name
+    """
+    for name_and_version_specifier in name_and_version_specifier_list:
+        # only accept a name without a version specifier
+        if object_name == name_and_version_specifier:
+            return True
+
+    return False 
+
+
+def has_compatible_version(name_and_version_specifier_list:list, object_name:str, object_version:str) -> bool:
+    """
+    returns True if the object_name matches an entry in the name_and_version_specifier_list that has no version,
+    or if the object_name matches and the object_version is compatible with the version specifier.
+    :param name_and_version_specifier_list: a list of names and (optional)version specifiers
+    :param object_name: the object name
+    :param object_version: the object version
+    """
+    try:
+        version = Version(object_version)
+    except InvalidVersion as e:
+        return False
```

Comment:
If the version string cannot be parsed, we should at least log a warning and the version string that can't be parsed